src/cart/crt_corpc.c

```diff
@@ -982,10 +982,17 @@ crt_corpc_req_hdlr(struct crt_rpc_priv *rpc_priv)
 		crt_corpc_reply_hdlr(&cb_info);
 	} else {
 		rc = crt_rpc_common_hdlr(rpc_priv);
-		if (rc != 0)
+		if (rc != 0) {
 			RPC_ERROR(rpc_priv,
 				  "crt_rpc_common_hdlr failed, rc: %d\n",
 				  rc);
+			/* Don't return error immediately since we already
+			 * sent rpcs to all children and need to wait for
+			 * their response before we call completion cb
+			 */
+			co_info->co_rc = rc;
+			rc = 0;
+		}
```

Comment:
Do we need to go through some function to check if there is any child RPC inflight? E.g., co_info->co_child_num, co_info->co_local_done, etc.

Reply:
I am not sure what we would do differently whether we have child rpcs in flight or not.
If we have children in flight, it will wait for children via aggregation like it does in case of success. Eventually in either case it should execute (in crt_corpc_complete()):

 580         am_root = (myrank == co_info->co_root);
 581         if (am_root) {
 582                 crt_rpc_complete(rpc_priv, co_info->co_rc);
 583                 RPC_DECREF(rpc_priv); /* destroy */
 584         } else {
 585                 if (co_info->co_rc != 0)
 586                         crt_corpc_fail_parent_rpc(rpc_priv, co_info->co_rc);

So it will either call local completion callback with the co_rc that we stored or it will propagate that rc to the parent.

Reply:
looks like should check crp_reply_pending flag, if it is set then the reply is not sent out this change will cause the reply aggregation cannot finish (then hang or timeout maybe).

Another kind of fixing could be that don't change the code here (still return error same as before), but seems should fix some refcount cases like  crt_rpc_handler_common() -> crt_corpc_common_hdlr() -> crt_corpc_initiate() -> crt_corpc_req_hdlr() failed, then in crt_rpc_handler_common it will RPC_DECREF(rpc_priv) that possibly unsafe as it still need to handle the reply aggregation.

I'm not sure what test case caused that so above is just FYI.

Reply:
I will add clearing of reply_pending flag here.
As to the case we are trying to fix is:

crt_req_send()->crt_corpc_req_hdlr() -> on error call crt_rpc_complete().
So if a local handler returned error, we will issue crt_rpc_complete() on root, while still having rpcs go out to children. This then can cause children to get corrupted data in input since caller can now free input data after corpc completion callback.

Reply:
> I will add clearing of reply_pending flag here.

only clearing the flag seems not enough, in that case the local handler failed before sending local reply so the aggregation possibly cannot work as expected, probably need to add a co_child_ack_num (and if it already matched with co_child_num then need to finished the corpc).


> As to the case we are trying to fix is:
> 
> crt_req_send()->crt_corpc_req_hdlr() -> on error call crt_rpc_complete().
> So if a local handler returned error, we will issue crt_rpc_complete() on root, while still having rpcs go out to children. This then can cause children to get corrupted data in input since caller can now free input data after corpc completion callback.

I just checked that in crt_req_send with the code of:
out:
	/* internally destroy the req when failed */
	if (rc != 0) {
		if (!rpc_priv->crp_coll) {
			crt_rpc_complete(rpc_priv, rc);
			/* failure already reported through complete cb */
			if (complete_cb != NULL)
				rc = 0;
		}
		RPC_DECREF(rpc_priv);
	}
	/* corresponds to RPC_ADDREF in this function */
	RPC_DECREF(rpc_priv);
	return rc;

so for corpc the crt_rpc_complete() is not called, the problem maybe due to the decref.

Reply:
thanks for checking. indeed, i missed the part where crt_rpc_complete() will only be called for non collective calls; i ll go through ref counts and see whats going on there

Reply:
welcome, BTW, even if you fix the refcount issue, here when the local handler failed, seems still need to add some err handling to fix the case that local reply is not sent then cause the reply aggregation cannot work as expected.

Reply:
If the crt_corpc_common_hdlr() fails then we already do crt_hg_reply_error_send() which should also now cover the case of local handler failing